### ISSUE TITLE
Inconsistent results from `jax.random.categorical` with `bfloat16` vs `float32` probabilities

### ISSUE THREAD

**Issue by spktrm** [jax-ml/jax#30879]
### Description

Noticed that `jax.random.categorical` gives noticeably different results when sampling from the same logits in bfloat16 vs float32. This is especially relevant for LLM output heads, where logits are often stored in reduced precision to save memory/bandwidth. Even small rounding changes near decision boundaries can end up picking a different token and changing beam/top-p/top-k search paths. This is more noticeable where `dims > 1000`.

```python
import numpy as np
import jax
import jax.numpy as jnp

num = 100
dims = 10000
key = jax.random.key(42)

arr = jax.random.uniform(key, (dims,))

unique_16 = jnp.unique(
    jax.random.categorical(key, arr.astype(jnp.bfloat16), shape=(num,)),
    return_counts=True
)

unique_32 = jnp.unique(
    jax.random.categorical(key, arr.astype(jnp.float32), shape=(num,)),
    return_counts=True
)

set_16 = set(np.array(unique_16[0]).tolist())
set_32 = set(np.array(unique_32[0]).tolist())

assert set_32 - set_16 == 0 # Rarely true if at all for different seeds
```

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.7.0
jaxlib: 0.7.0
numpy:  2.0.2
python: 3.11.13 (main, Jun  4 2025, 08:57:29) [GCC 11.4.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='dd07cc57c10d', release='6.1.123+', version='#1 SMP PREEMPT_DYNAMIC Sun Mar 30 16:01:29 UTC 2025', machine='x86_64')

**Comment by jakevdp** [comment]
Hi - thanks for the question! This looks similar to the question in #27002.

In short, this is behaving as expected: the random sampler will sample differently depending on the precision being used. For float32 categories, the samples are generated with 32 bits of randomness, and for bfloat16 categories, the samples are generated with 16 bits of randomness. This leads to different uniform samples for a given key:
```python
import jax
key = jax.random.key(0)
print(jax.random.uniform(key, (5,), dtype='float32'))
print(jax.random.uniform(key, (5,), dtype='bfloat16'))
```
```
[0.947667   0.9785799  0.33229148 0.46866846 0.5698887 ]
[0.648438 0.570312 0.15625 0.203125 0.710938]
```
Because categorical samples are constructed from uniform samples, you'll also see different categorical samples depending on dtype. In both cases the samples should follow the correct distribution (please report the bug if they don't!) but with different dtypes you will get different draws from the distribution.

I hope that makes sense!

**Comment by jakevdp** [comment]
Closing, as this is working as intended.
